src/dict.js

```diff
@@ -69,68 +99,66 @@ Sk.builtin.dict.tp$call = function(args, kw) {
 Sk.abstr.setUpInheritance("dict", Sk.builtin.dict, Sk.builtin.object);
 Sk.abstr.markUnhashable(Sk.builtin.dict);
 
-var kf = Sk.builtin.hash;
+var reg = /^[0-9!#_]/; // avoid clashes with complex key hashes 
+// and the problem with js ints not preserving order: str('1') => "!1"
 
-Sk.builtin.dict.prototype.key$lookup = function (bucket, key) {
-    var item;
-    var eq;
-    var i;
-
-    // Fast path: We spend a *lot* of time looking up strings
-    // in dictionaries. (Every attribute access, for starters.)
-    if (key.ob$type === Sk.builtin.str) {
-        for (i = 0; i < bucket.items.length; i++) {
-            item = bucket.items[i];
-            if (item.lhs.ob$type === Sk.builtin.str && item.lhs.v === key.v) {
-                return item;
-            }
-        }
-        return null;
-    }
-
-    for (i = 0; i < bucket.items.length; i++) {
-        item = bucket.items[i];
-        eq = Sk.misceval.richCompareBool(item.lhs, key, "Eq");
-        if (eq) {
-            return item;
-        }
+function kf(key) {
+    // str => jsstr().replace(/^[0-9!#_]/, "!$&") avoids conflicts
+    // other => hash.v value from builtin.hash (javascript number)
+    let key_hash = key.$savedKeyHash_; 
+    if (key_hash !== undefined) {
+        return key_hash;
+    } else if (key.ob$type === Sk.builtin.str) {
+        key_hash = key.$jsstr().replace(reg, "!$&"); 
+        key.$savedKeyHash_ = key_hash;
+        return key_hash;
+    } else if (typeof key === "string") {
+        // temporary while sysModules allows javascript strings as keys to python dicts
+        return key.replace(reg, "!$&");
     }
-
-    return null;
+    key_hash = Sk.builtin.hash(key).v; // builtin.hash returns an int
+    key.$savedKeyHash_ = key_hash;
+    return key_hash;
 };
 
-Sk.builtin.dict.prototype.key$pop = function (bucket, key) {
-    var item;
-    var eq;
-    var i;
+Sk.builtin.dict.prototype.sk$asarray = function () {
+    return Object.values(this.entries).map((x) => x.lhs);
+};
 
-    for (i = 0; i < bucket.items.length; i++) {
-        item = bucket.items[i];
-        eq = Sk.misceval.richCompareBool(item.lhs, key, "Eq");
-        if (eq) {
-            bucket.items.splice(i, 1);
-            this.size -= 1;
+Sk.builtin.dict.prototype.get$bucket_item = function (key, hash_value) {
+    // slow path to get an item, check the bucket with specific hash_value until we find a match
+    const bucket = this.buckets[hash_value];
+    let bucket_key, item;
+    if (bucket === undefined) {
+        return;
+    }
+    for (let i = 0; i < bucket.length; i++) {
+        item = bucket[i];
+        if (item === undefined) {
+            // free slot from having deleted an item.
+            continue;
```

Comment:
given the empty slots now...
If I instead do `for (let item of bucket)` it will just skip these slots. better?